nf_core/download.py

```diff
@@ -1335,3 +1384,105 @@ def bare_clone(self, destination):
                 self.repo.clone(os.path.abspath(destination), bare=True)
             except (OSError, GitCommandError, InvalidGitRepositoryError) as e:
                 log.error(f"[red]Failure to create the pipeline download[/]\n{e}\n")
+
+
+# Distinct errors for the container download, required for acting on the exceptions
+
+
+class ContainerError(Exception):
+    """A class of errors related to pulling containers with Singularity/Apptainer"""
+
+    def __init__(self, container, registry, address, out_path, singularity_command, error_msg):
+        self.container = container
+        self.registry = registry
+        self.address = address
+        self.out_path = out_path
+        self.singularity_command = singularity_command
+        self.error_msg = error_msg
+
+        for line in error_msg:
+            if re.search(r"dial\stcp.*no\ssuch\shost", line):
+                self.error_type = self.RegistryNotFound(self)
+                break
+            elif (
+                re.search(r"requested\saccess\sto\sthe\sresource\sis\sdenied", line)
+                or re.search(r"StatusCode:\s404", line)
+                or re.search(r"invalid\sstatus\scode\sfrom\sregistry\s400", line)
+            ):
+                # Unfortunately, every registry seems to return an individual error here:
+                # Docker.io: denied: requested access to the resource is denied
+                #                    unauthorized: authentication required
+                # Quay.io: StatusCode: 404,  <!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2 Final//EN">\n']
+                # ghcr.io: Requesting bearer token: invalid status code from registry 400 (Bad Request)
```

Comment:
No, the actual problem is that all the error handling is just done by `"parsing"` (actually just running through regexes) whatever singularity prints to stdout/stderr in response to trying `singularity pull` with the container path. 
This was acceptable, because in the previous version, scanning for the keyword "FATAL" and then killing the command was all that was needed. 

Now, for the first time, it was required to distinguish different errors based on what was returned. I was/am not really happy with that solution, because I anticipated that this would be very flaky (e.g. because the wording changes with different versions of Singularity/Apptainer or the error messages are printed in a different language due to localization) but also had no better idea how to address this. 

Indeed, avoiding `singularity pull` entirely and testing the path directly from Python now seems like an excellent idea to me. However, I fear that the Python request library just handles http? I think, this might be a different protocol and thus the 404 code returned by Quay.io is not standard?